modules/dists/SparseBlockDist.chpl

```diff
@@ -117,6 +118,64 @@ class SparseBlockDom: BaseSparseDom {
     dsiAdd((ind,));
   }
 
+  proc dsiBulkAdd(inds: [] index(rank, idxType), isSorted=false, isUnique=false,
+      preserveInds=true){
+  
+    if !isSorted && preserveInds {
+      var _inds = inds;
+      bulkAdd_help(_inds, isSorted, isUnique); 
+    }
+    else {
+      bulkAdd_help(inds, isSorted, isUnique);
+    }
+  }
+
+  proc bulkAdd_help(inds: [] index(rank,idxType), isSorted=false, isUnique=false) {
+
+    if !isSorted then
+      QuickSort(inds);
+
+    var localIndexes: [dist.targetLocDom] [inds.domain] index(rank,idxType);
+    var localIndexCnts: [dist.targetLocDom] int =
+      inds.domain.low;
+    if bulkAddMemoize {
+      var lastLocaleIndexCache = dist.targetLocDom.low;
+      for i in inds {
+        if locDoms[lastLocaleIndexCache].parentDom.member(i) {
+          localIndexes[lastLocaleIndexCache][localIndexCnts[lastLocaleIndexCache]] = i;
+          localIndexCnts[lastLocaleIndexCache] += 1;
+        }
+        else {
+          for localeIndex in dist.targetLocDom {
```

Comment:
The part I do not understand is why we need a for loop here at all. Can't we sort the locales by destination?